arangod/Aql/Query.h

```diff
@@ -367,7 +367,7 @@ class Query {
 
   /// @brief set of DataSources used in the query
   ///        needed for the query cache, value LogicalDataSource::system()
-  std::unordered_set<std::shared_ptr<arangodb::LogicalDataSource>> _queryDataSources;
+  std::vector<std::weak_ptr<arangodb::LogicalDataSource>> _queryDataSources;
```

Comment:
I would leave `shared_ptr` for `Query` since query and underlying transaction own data sources involved